File: android/app/src/main/java/app/rentivo/domain/Identifiers.kt
package app.rentivo.domain

import java.io.File

/**
 * Opaque server identifiers.
 *
 * Swift models these as one generic `ResourceID<Tag>` with phantom tags; Kotlin has no zero-cost
 * equivalent, so each identifier is its own inline value class over the same `rawValue` String.
 * The values stay opaque — never parse them as UUIDs.
 */
@JvmInline value class BillingID(val rawValue: String)

@JvmInline value class BillID(val rawValue: String)

@JvmInline value class BillingItemID(val rawValue: String)

@JvmInline value class BillLineItemID(val rawValue: String)

@JvmInline value class ReceiptID(val rawValue: String)

@JvmInline value class ExpenseID(val rawValue: String)

@JvmInline value class AttachmentID(val rawValue: String)

@JvmInline value class OrganizationID(val rawValue: String)

@JvmInline value class InvitationID(val rawValue: String)

@JvmInline value class APIKeyID(val rawValue: String)

@JvmInline value class PasskeyID(val rawValue: String)

@JvmInline value class CommunicationID(val rawValue: String)

@JvmInline value class RecipientID(val rawValue: String)

@JvmInline
value class WorkspaceID(val rawValue: String) {
  companion object {
    val personal = WorkspaceID(rawValue = "personal")
  }
}

/** Bytes staged for a multipart upload. */
class FileUpload(
  val data: ByteArray,
  val filename: String,
  val mediaType: String,
) {
  val byteCount: Int get() = data.size

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is FileUpload) return false
    return data.contentEquals(other.data) &&
      filename == other.filename &&
      mediaType == other.mediaType
  }

  override fun hashCode(): Int {
    var result = data.contentHashCode()
    result = 31 * result + filename.hashCode()
    result = 31 * result + mediaType.hashCode()
    return result
  }

  override fun toString(): String =
    "FileUpload(filename=$filename, mediaType=$mediaType, byteCount=$byteCount)"
}

object AttachmentUploadRules {
  /** Mirrors `ALLOWED_ATTACHMENT_TYPES` in the backend billing attachment model. */
  val allowedMediaTypes: Set<String> = setOf("application/pdf", "image/jpeg", "image/png")

  /** Mirrors `MAX_ATTACHMENT_SIZE` in the backend billing attachment model. */
  const val maximumByteCount: Int = 10 * 1024 * 1024

  fun validated(upload: FileUpload): FileUpload {
    val mediaType = upload.mediaType.substringBefore(';').trim().lowercase()
    if (mediaType !in allowedMediaTypes) throw DemoError("Envie um arquivo PDF, JPEG ou PNG.")
    if (upload.byteCount == 0) throw DemoError("O arquivo selecionado está vazio.")
    if (upload.byteCount > maximumByteCount) {
      throw DemoError("O arquivo excede o limite de 10 MB.")
    }
    return FileUpload(data = upload.data, filename = upload.filename, mediaType = mediaType)
  }
}

/** A file the API layer wrote into the downloads directory owned by `DownloadedFileStore`. */
data class DownloadedFile(
  val file: File,
  val filename: String,
  val mediaType: String,
)

File: ios/Rentivo/Domain/Identifiers.swift
import Foundation
import UniformTypeIdentifiers

public struct ResourceID<Tag>: RawRepresentable, Hashable, Codable, Sendable, Identifiable {
  public let rawValue: String

  public var id: String { rawValue }

  public init(rawValue: String) {
    self.rawValue = rawValue
  }
}

public enum BillingIDTag: Sendable {}
public typealias BillingID = ResourceID<BillingIDTag>
public enum BillIDTag: Sendable {}
public typealias BillID = ResourceID<BillIDTag>
public enum BillingItemIDTag: Sendable {}
public typealias BillingItemID = ResourceID<BillingItemIDTag>
public enum BillLineItemIDTag: Sendable {}
public typealias BillLineItemID = ResourceID<BillLineItemIDTag>
public enum ReceiptIDTag: Sendable {}
public typealias ReceiptID = ResourceID<ReceiptIDTag>
public enum ExpenseIDTag: Sendable {}
public typealias ExpenseID = ResourceID<ExpenseIDTag>
public enum AttachmentIDTag: Sendable {}
public typealias AttachmentID = ResourceID<AttachmentIDTag>
public enum OrganizationIDTag: Sendable {}
public typealias OrganizationID = ResourceID<OrganizationIDTag>
public enum InvitationIDTag: Sendable {}
public typealias InvitationID = ResourceID<InvitationIDTag>
public enum APIKeyIDTag: Sendable {}
public typealias APIKeyID = ResourceID<APIKeyIDTag>
public enum PasskeyIDTag: Sendable {}
public typealias PasskeyID = ResourceID<PasskeyIDTag>
public enum CommunicationIDTag: Sendable {}
public typealias CommunicationID = ResourceID<CommunicationIDTag>
public enum RecipientIDTag: Sendable {}
public typealias RecipientID = ResourceID<RecipientIDTag>
public enum WorkspaceIDTag: Sendable {}
public typealias WorkspaceID = ResourceID<WorkspaceIDTag>

public extension ResourceID where Tag == WorkspaceIDTag {
  static let personal = Self(rawValue: "personal")
}

public struct FileUpload: Hashable, Sendable {
  public let data: Data
  public let filename: String
  public let mediaType: String

  public init(data: Data, filename: String, mediaType: String) {
    self.data = data
    self.filename = filename
    self.mediaType = mediaType
  }

  public var byteCount: Int { data.count }

  public static func from(url: URL, policy: FileUploadPolicy? = nil) throws -> Self {
    let filename = url.lastPathComponent
    let mediaType = UTType(filenameExtension: url.pathExtension)?.preferredMIMEType
      ?? "application/octet-stream"
    if policy == .rentivoDocument {
      guard RentivoUploadPolicy.allowedMediaTypes.contains(mediaType.lowercased()) else {
        throw FileUploadValidationError.unsupportedType
      }
      let values = try url.resourceValues(forKeys: [.fileSizeKey])
      guard let fileSize = values.fileSize, fileSize <= RentivoUploadPolicy.maxByteCount else {
        throw FileUploadValidationError.tooLarge
      }
    }
    return try Self(data: Data(contentsOf: url), filename: filename, mediaType: mediaType)
  }

  /// Claims a document-picker URL and performs its metadata/data reads off the UI actor.
  public nonisolated static func fromSecurityScoped(
    url: URL, policy: FileUploadPolicy? = nil
  ) async throws -> Self {
    let accessGranted = url.startAccessingSecurityScopedResource()
    defer { if accessGranted { url.stopAccessingSecurityScopedResource() } }
    return try from(url: url, policy: policy)
  }
}

public enum FileUploadPolicy: Hashable, Sendable {
  case rentivoDocument
}

public enum RentivoUploadPolicy {
  public static let maxByteCount = 10 * 1024 * 1024
  public static let allowedMediaTypes: Set<String> = [
    "application/pdf", "image/jpeg", "image/png",
  ]
}

public enum FileUploadValidationError: Error, Equatable, Sendable {
  case unsupportedType
  case tooLarge
}

public enum AttachmentUploadRules {
  /// Mirrors `ALLOWED_ATTACHMENT_TYPES` in `backend/rentivo/models/billing_attachment.py`.
  public static let allowedMediaTypes: Set<String> = [
    "application/pdf", "image/jpeg", "image/png",
  ]

  /// Mirrors `MAX_ATTACHMENT_SIZE` in `backend/rentivo/models/billing_attachment.py`.
  public static let maxByteCount = 10 * 1024 * 1024

  public static func validated(_ upload: FileUpload) throws -> FileUpload {
    let mediaType = upload.mediaType
      .split(separator: ";", maxSplits: 1)
      .first?
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .lowercased() ?? ""
    guard allowedMediaTypes.contains(mediaType) else {
      throw DemoError(message: "Envie um arquivo PDF, JPEG ou PNG.")
    }
    guard upload.byteCount > 0 else {
      throw DemoError(message: "O arquivo selecionado está vazio.")
    }
    guard upload.byteCount <= maxByteCount else {
      throw DemoError(message: "O arquivo excede o limite de 10 MB.")
    }
    return FileUpload(data: upload.data, filename: upload.filename, mediaType: mediaType)
  }
}

public struct DownloadedFile: Hashable, Sendable {
  public let fileURL: URL
  public let filename: String
  public let mediaType: String

  public init(fileURL: URL, filename: String, mediaType: String) {
    self.fileURL = fileURL
    self.filename = filename
    self.mediaType = mediaType
  }
}

extension DownloadedFile: Identifiable {
  public var id: URL { fileURL }
}
